### ISSUE TITLE
500+ multi-labels only predicts zeros

### ISSUE THREAD

**Issue by irdeen** [kaushaltrivedi/fast-bert#52]
I now have a dataset of 500+ labels. However, at first fast-bert predicts various values between 0-1 for every label, but the more I train the more it predicts zeros for everything. Logically, it seems wise as only 1/500 is a positive label while the rest are zeros. Is there a way to fix this? Can I change the loss function somehow? Possibly introduce class weights or something to really penalize false-negatives?

**Comment by guy-mor** [comment]
You could use the pos_weight argument to the loss function (BCEWithLogitsLoss or BCELoss) to give greater weight to to positive instances than to negative instances, increasing recall at the expense of precision.

**Comment by irdeen** [comment]
Ah that's great! Although pos_weight seems to be a fixed vector you set from the start to boost up certain label weights if you have an imbalanced dataset. What I want is an adaptive pos_weight that heavily penalizes the loss as soon as fast-bert predicts a 0 instead of a 1 for a label. I don't know if that's possible though.

**Comment by guy-mor** [comment]
pos_weight differs from regular class weights, and I think that it should do exactly what you want it to do. You have very sparse labels where for each sample the label mostly contains 0s, and one (or a few?) 1s, so naturally the model learns to guess 0 all the time, getting good (binary) accuracy but poor recall. pos_weight will weight up the loss so that predictions in the label whose ground truth equals one count more than those that equal 0 - so that that the model would care more about getting the few positive labels right, then getting the many negative labels right. 

Finding the correct value for pos_weight might require some experimentation, though. If in addition to multilabel sparsity you also have class imbalance (some classes are equal 1 more often than others) than it's worth reflecting this in the pos_weight vector. But it's still a question by how much the positive values should be up-weighted.

BTW, I had a similar problem and pos_weight was able to solve it for a reduced problem set (if I only used 200 labels instead of 600+) but was unable yet to find a good values that make working with 600+ labels work. Let me know if you manage this.

**Comment by irdeen** [comment]
Oh I see! My bad for misunderstanding, very grateful for your help! You're precisely right on what I'm trying to achieve. I'm not quite sure how I would set up this pos_weight parameter though. Is it a vector? How do I calculate it? I'm still a bit confused even after googling up on it.

**Comment by guy-mor** [comment]
It's a vector the size of your label, but can also be a scalar if you want the same positive weight for all categories in the label (scalar should simply be broadcast in the calculation). I've seen others use a default value of 10.0 (has to be float), log(num_categories), or number_neg_samples / number_pos_samples (for a vector of varying weights, if you also have class imbalance). It's worth playing around with.

**Comment by irdeen** [comment]
Ah that makes sense. Hmm doesn't seem to work as I notice no difference in the loss when I add the pos_weight parameter, as such: `BCEWithLogitsLoss(pos_weight=torch.tensor(np.ones(670)*10.0))`. This line does run, as I get errors if I provide another label length other than 670, it's just that nothing seems to happen. Any ideas what could be wrong?

Example of the data:
```
"Spare Parts for Commissioning and start up each item", 0,0, ... ,1, ... ,0,0 (670 labels)
"100% Project Specification HS84218-00-PIP-SPE-0007-", 0,0, ... ,1, ... ,0,0 (670 labels)
...
```

Printout of the training process:
```
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 32768.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 16384.0-------------------------------| 0.15% [1/689 00:00<08:33]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 8192.0--------------------------------| 0.29% [2/689 00:01<07:53]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 4096.0--------------------------------| 0.44% [3/689 00:02<07:39]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 2048.0--------------------------------| 0.58% [4/689 00:02<07:31]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 1024.0--------------------------------| 0.73% [5/689 00:03<07:27]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 512.0---------------------------------| 0.87% [6/689 00:03<07:24]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 256.0---------------------------------| 1.02% [7/689 00:04<07:22]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 128.0---------------------------------| 1.16% [8/689 00:05<07:20]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 64.0----------------------------------| 1.31% [9/689 00:05<07:19]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 32.0----------------------------------| 1.45% [10/689 00:06<07:17]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 16.0----------------------------------| 1.60% [11/689 00:07<07:16]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 8.0-----------------------------------| 1.74% [12/689 00:07<07:15]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 4.0-----------------------------------| 1.89% [13/689 00:08<07:14]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 2.0-----------------------------------| 2.03% [14/689 00:08<07:13]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 1.0-----------------------------------| 2.18% [15/689 00:09<07:12]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 0.5-----------------------------------| 2.32% [16/689 00:10<07:11]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 0.25----------------------------------| 2.47% [17/689 00:10<07:10]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 0.125---------------------------------| 2.61% [18/689 00:11<07:09]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 0.0625--------------------------------| 2.76% [19/689 00:12<07:08]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 0.03125-------------------------------| 2.90% [20/689 00:12<07:08]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 0.015625------------------------------| 3.05% [21/689 00:13<07:07]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 0.0078125-----------------------------| 3.19% [22/689 00:14<07:06]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 0.00390625----------------------------| 3.34% [23/689 00:14<07:05]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 0.001953125---------------------------| 3.48% [24/689 00:15<07:04]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 0.0009765625--------------------------| 3.63% [25/689 00:15<07:04]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 0.00048828125-------------------------| 3.77% [26/689 00:16<07:03]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 0.000244140625------------------------| 3.92% [27/689 00:17<07:02]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 0.0001220703125-----------------------| 4.06% [28/689 00:17<07:02]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 6.103515625e-05-----------------------| 4.21% [29/689 00:18<07:01]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 3.0517578125e-05----------------------| 4.35% [30/689 00:19<07:00]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 1.52587890625e-05---------------------| 4.50% [31/689 00:19<06:59]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 7.62939453125e-06---------------------| 4.64% [32/689 00:20<06:59]
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 3.814697265625e-06--------------------| 4.79% [33/689 00:21<06:58]
```
This loss continues to dive pretty fast down to 10^-200, which seems really strange. Maybe it's something else I'm missing?

**Comment by guy-mor** [comment]
I'm not sure, but it could very well be that 10 isn't a significant enough weight for the positive values since 0s outnumber 1s so much in your data. I suggest trying different, higher values. If this experimentation is too computationally costly perhaps try to overfit a smaller sample of the training data. PS instead of log(num_categories) I should have written sqrt(num_categories), which in your case is substantially higher than the former. But if I understand your data correctly number_neg_samples / number_pos_samples should produce even higher weights.

**Comment by mmortazavi** [comment]
I am observing very similar issues. Multi-label classification with 5 labels. Each data point has a minimum of 1 label and maximum of 2 or 3 labels (encoded as 1s). It is not that sparse like 600+ labels in the case of @irdeen. I see that the predictor acts as it is binary. It returns two classes with high probability and rest are rather close to zero regardless of the number of  labels. The label distribution is skewed too (two labels 500 samples, others are around 80), so I have a imbalance dataset as @guy-mor mentioned. 

The question is about the `pos_weight` argument. Do we have this `pos_weight `argument to the loss function (`BCEWithLogitsLoss `or `BCELoss`) exposed directly in the `learner ` object in `fast-bert`? Or you guys are talking modifying the loss function in the source code? There are no documentations and the conversations above were not clear. 

And would you @irdeen explain the `BCEWithLogitsLoss(pos_weight=torch.tensor(np.ones(670)*10.0)) ` in your case. You have 670 labels. Doesn't `torch.tensor(np.ones(670)*10.0)` construct a tensor of size 670 with all 10.0 values? How does this work as a `pos_weight`? All your labels are equally weighted, aren't they? I thought the point is to weigh according to their counts or distributions. Your feedback is appreciated.
